SDK Integration: OpenAI Agents  + Client .dxt (#146)
kluster-ai/docs: get-started/mcp/sdk-integrations/openai-agents.md

---

> title: Integrate OpenAI Agents SDK with kluster Verify

Reviewer: This title is a bit long, but I'm going to let it go and see if we get flagged for it in our semrush site audit 🤷‍♀️

> description: This guide walks you through integrating OpenAI Agents SDK with kluster Verify to build reliable AI agents with real-time hallucination detection and fact-checking capabilities.

Reviewer: Meta descriptions should be between 120-160 characters.

---

# Integrate OpenAI Agents SDK with kluster Verify

[OpenAI Agents SDK](https://openai.github.io/openai-agents-python/){target=_blank} provides a powerful framework for building AI agents that can use tools, maintain context, and interact with external services. By integrating with [kluster Verify](/get-started/verify/overview/){target=_blank}, you can create agents that not only generate responses using [kluster.ai's](https://www.kluster.ai/){target=_blank} language models but also perform real-time reliability checks to detect hallucinations and validate factual claims with internet-sourced verification.

> This guide demonstrates how to integrate the `Agent` and `Runner` classes from the OpenAI Agents SDK with kluster.ai's API and MCP server, then walks through building an interactive chatbot with kluster Verify's hallucination detection capabilities.

Reviewer: ```suggestion
This guide demonstrates how to integrate the `Agent` and `Runner` classes from the OpenAI Agents SDK with kluster.ai's API and MCP server, and then walks through building an interactive chatbot that utilizes kluster Verify's hallucination detection capabilities.
```

## Prerequisites

--8<-- 'text/kluster-api-onboarding.md'

> - **Python 3.9+** with pip installed.

Reviewer: ```suggestion
- **Python 3.9+**: Make sure pip is also installed.
```

> - **[A Python virtual environment](https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/){target=_blank}** - this is optional but recommended. Ensure that you enter the Python virtual environment before following along with this tutorial.

Reviewer: ```suggestion
- **[A Python virtual environment](https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/){target=_blank}**: This is optional but recommended. Ensure that you enter the Python virtual environment before following along with this tutorial.
```

> - **OpenAI Agents SDK packages installed** - install the [`openai-agents` packages](https://github.com/openai/openai-agents-python){target=_blank}:

Reviewer: ```suggestion
- **OpenAI Agents SDK packages installed**: Use the following command to install the [`openai-agents` packages](https://github.com/openai/openai-agents-python){target=_blank}.
```

    ```bash
    pip install "openai>=1.93.0" "openai-agents>=0.1.0" "mcp[cli]>=1.10.1"
    ```

> - **MCP enabled**: Via [platform](https://platform.kluster.ai){target=_blank} or [API](/get-started/mcp/cloud/api/).

Reviewer: ```suggestion
- **MCP enabled**: This can be done via the [platform](https://platform.kluster.ai){target=_blank} or the [API](/get-started/mcp/cloud/api/){target=\_blank}.
```

## Quick start

It's easy to integrate kluster Verify with OpenAI Agents SDK—when configuring the agent, point your `AsyncOpenAI` instance to the correct base URL and configure the following settings:

  - **Base URL**: Use `https://api.kluster.ai/v1` to send requests to the kluster.ai endpoint.
  - **API key**: Replace with your kluster.ai API key. If you don't have one yet, refer to the [Get an API key guide](/get-started/get-api-key/).

>   - **Select your model**: Choose a model with tool support from [kluster.ai's models](https://platform.kluster.ai/models){target=_blank} (filter by "Tool Support").

Reviewer: Wdym filter by Tool Support? I don't see any filtering as an option aside from the top menu you can switch between

<img width="726" height="331" alt="Screenshot 2025-07-17 at 12 30 39 AM" src="https://github.com/user-attachments/assets/b80a96a3-4c56-4269-a6ba-4afb7b4ce323" />

Author: <img width="1616" height="669" alt="image" src="https://github.com/user-attachments/assets/c193672f-ff0d-47db-a946-f1ad218af526" />

it's not yet updated on docs, but already on platform.

Reviewer: I don't see it when I'm on the platform site - is your screenshot of a staging site or something?

<img width="1728" height="1042" alt="Screenshot 2025-07-17 at 2 50 47 PM" src="https://github.com/user-attachments/assets/2ac76649-f4fe-49e8-a342-42d737855725" />

Reviewer: Ahhhh haha when I zoom out to like 90% I can see it. The table isn't scrollable so it just cuts it off

  - **MCP server**: Configure the MCP server URL and token for kluster Verify's reliability checking.

```python

> from agents import Agent, Runner, set_tracing_disabled, set_default_openai_api, OpenAIChatCompletionsModel

Reviewer: Has this been formatted? Some of these lines are pretty long

        model="deepseek-ai/DeepSeek-V3-0324",
        openai_client=kluster_client
    ),
    mcp_servers=[mcp_server]
)

# Run the agent
async def main():
    await mcp_server.connect()
    result = await Runner.run(agent, "Is the Earth flat?")
    print(result.final_output)

asyncio.run(main())
```

That's all you need to start with OpenAI Agents SDK and kluster Verify! Next, this guide will explore building an interactive chatbot that showcases kluster Verify's hallucination detection and real-time verification capabilities.

!!! info "Model requirement"
    Only models with **tool support** can use MCP verification. Filter models by "Tool Support" on the [platform models page](https://platform.kluster.ai/models){target=_blank} to see compatible options.

!!! warning "Known async cleanup issue"
    You may see error messages about "Exception ignored in atexit callback" when the script exits. This is a [known issue](https://github.com/modelcontextprotocol/python-sdk/issues/521){target=_blank} with MCP's asyncio cleanup that **does not affect functionality**. Your script will work correctly despite these messages.

!!! info "Self-hosted option"

>     For self-hosted MCP, use `http://localhost:3001/stream` and your kluster.ai API key.

Reviewer: When there are too many admonitions (and colors) it's like your eyes are getting pulled to too many things at once and it does more harm than good.

I would remove the info admos, keep the warning admo.

```
Only models with tool support can use MCP verification. To find compatible options, filter by "Tool Support" on the [platform models page](https://platform.kluster.ai/models){target=_blank}.

For self-hosted MCP, use `http://localhost:3001/stream` along with your kluster.ai API key.

!!! warning "Known async cleanup issue"
    You may see error messages about "Exception ignored in atexit callback" when the script exits. This is a [known issue](https://github.com/modelcontextprotocol/python-sdk/issues/521){target=_blank} with MCP's asyncio cleanup that **does not affect functionality**. Your script will work correctly despite these messages.
```

## Enable MCP

If you prefer, you can enable MCP and obtain your token via API calls instead of using the platform interface. This approach allows you to automate the token retrieval process and integrate it directly into your application setup workflow.

```python
import requests

# Enable MCP
response = requests.post(
    "https://api.kluster.ai/v1/mcp/enable",
    headers={"Authorization": f"Bearer {api_key}"}
)

# Get status with token
status = requests.get(
    "https://api.kluster.ai/v1/mcp/status", 
    headers={"Authorization": f"Bearer {api_key}"}
)
mcp_token = status.json()["apiKey"]
```

## Build an interactive chatbot

> This example creates an interactive chatbot where you can ask questions and see kluster Verify validate the responses in real-time, demonstrating how verification enhances conversational AI.

Reviewer: This is a pretty big code block you're dropping here. Nothing needs explanation?

```python

> from agents import Agent, Runner, set_tracing_disabled, set_default_openai_api, OpenAIChatCompletionsModel

Reviewer: Too long; needs formatting.

    base_url="https://api.kluster.ai/v1",
    api_key=api_key
)
# Create MCP server
mcp_server = MCPServerStreamableHttp(
    params={
        "url": "https://api.kluster.ai/v1/mcp",
        "headers": {"Authorization": f"Bearer {mcp_token}"},
        "timeout": 15,
        "sse_read_timeout": 15
    },
    client_session_timeout_seconds=30  # Increase from default 5 seconds to 30 seconds
)

# Create chatbot agent
agent = Agent(
    name="VerifyChatbot",
    instructions="Use kluster Verify to validate factual claims and provide reliable responses.",
    model=OpenAIChatCompletionsModel(
        model="deepseek-ai/DeepSeek-V3-0324",
        openai_client=kluster_client
    ),
    mcp_servers=[mcp_server]
)

# Interactive chat loop
async def main():
    await mcp_server.connect()
    
    print("🤖 Chatbot ready! Type 'quit' to exit.")
    
    while True:
        user_input = input("\n👤 You: ").strip()
        
        if user_input.lower() in ['quit', 'exit', 'q']:
            print("👋 Goodbye!")
            break
            
        if user_input:
            result = await Runner.run(agent, user_input)
            print(f"🤖 Bot: {result.final_output}")

asyncio.run(main())
```

> ## Complete example

Reviewer: A section is overkill here. You can just add this to the end of the above section

<details>

> <summary>Full integration script</summary>

Reviewer: Use an admo instead of html

```
??? code "Complete example"
    ... insert content...
```

        model=OpenAIChatCompletionsModel(
            model="deepseek-ai/DeepSeek-V3-0324",
            openai_client=kluster_client
        ),
        mcp_servers=[mcp_server]
    )
    
    await mcp_server.connect()
    
    print("\n✅ Chatbot ready! Type 'quit' to exit.")
    print("💬 Ask me anything and I'll verify my responses:\n")
    
    while True:
        try:
            # Get user input
            user_input = input("👤 You: ").strip()
            
            if user_input.lower() in ['quit', 'exit', 'q']:
                print("\n👋 Goodbye!")
                break
                
            if not user_input:
                continue
            
            # Get agent response with verification
            print("🤖 Bot: Thinking and verifying...")
            result = await Runner.run(agent, user_input)
            print(f"🤖 Bot: {result.final_output}\n")
            
        except KeyboardInterrupt:
            print("\n\n👋 Goodbye!")
            break
        except Exception as e:
            print(f"❌ Error: {e}\n")

if __name__ == "__main__":
    asyncio.run(main())
```
</details>

## Run the script

> ```bash
> python reliable_agent.py
> ```
> 
> Enter your credentials when prompted:
> 
> - kluster.ai API key (get one from [platform.kluster.ai](https://platform.kluster.ai){target=_blank}).

Reviewer: ```suggestion

1. Use the following command to run the script:

    ```bash
    python reliable_agent.py
    ```

2. Enter your kluster.ai API key when prompted. If you don't have one yet, refer to the [Get an API key guide](/get-started/get-api-key/){target=\_blank}.
```

> 

Reviewer: ```suggestion
```

Expected output:
```
🔑 Enter your kluster.ai API Key: 

✅ Chatbot ready! Type 'quit' to exit.
💬 Ask me anything and I'll verify my responses:

👤 You: Is it true that the Eiffel Tower was moved to London in May 2025?
🤖 Bot: Thinking and verifying...
🤖 Bot: No, the Eiffel Tower was not moved to London in May 2025 or at any other time. It remains in its original location in Paris, France. 

### Verification Details:
- **Explanation**: The search results confirm that the Eiffel Tower is located in Paris, and there is no credible information suggesting it was relocated to London. 
- **Supporting Sources**:
  - [Eiffel Tower - Wikipedia](https://en.wikipedia.org/wiki/Eiffel_Tower) describes its location as Paris, France.
  - Other sources mention proposals or ideas for towers in London but confirm these are unrelated to the Eiffel Tower.

The claim about the Eiffel Tower being moved to London is false.

👤 You: quit

👋 Goodbye!

> ```

Reviewer: Terminal output? Please use a termynal element

# Get credentials securely
api_key = getpass.getpass("Enter your kluster.ai API key: ")
mcp_token = getpass.getpass("Enter your MCP token: ")

# Create kluster.ai client
kluster_client = AsyncOpenAI(
    base_url="https://api.kluster.ai/v1",
    api_key=api_key
)

# Configure MCP server for verification
mcp_server = MCPServerStreamableHttp(
    params={
        "url": "https://api.kluster.ai/v1/mcp",
        "headers": {"Authorization": f"Bearer {mcp_token}"},
        "timeout": 15,
        "sse_read_timeout": 15
    },
    client_session_timeout_seconds=30  # Increase from default 5 seconds to 30 seconds
)

# Create agent with kluster Verify capabilities
agent = Agent(
    name="ReliableAgent",
    instructions="Use kluster Verify for factual claims. Verify can detect hallucinations and validate information against real-time sources.",
    model=OpenAIChatCompletionsModel(
        model="deepseek-ai/DeepSeek-V3-0324",
        openai_client=kluster_client
    ),
    mcp_servers=[mcp_server]
)

# Run the agent
async def main():
    await mcp_server.connect()
    result = await Runner.run(agent, "Is the Earth flat?")
    print(result.final_output)

asyncio.run(main())
```

> That's all needed to start with OpenAI Agents SDK and kluster Verify! Next, this guide will explore building an interactive chatbot that showcases kluster Verify's hallucination detection and real-time verification capabilities.

Reviewer 2: ```suggestion
That's all you need to start with OpenAI Agents SDK and kluster Verify! Next, this guide will explore building an interactive chatbot that showcases kluster Verify's hallucination detection and real-time verification capabilities.
```

> Only models with **tool support** can use MCP verification. To find compatible options, filter by "Tool Support" on the [platform models page](https://platform.kluster.ai/models){target=_blank}.

Reviewer: ```suggestion
Only models with tool support can use MCP verification. To find compatible options, filter by **Tool Support** on the [platform models page](https://platform.kluster.ai/models){target=_blank}.
```

UI elements should be in bold, no quotes needed.
